### PR TITLE
[REEF-642]:Remove custom assembly loading logic

### PR DESCRIPTION
**Pull request by beysims**
This addressed the issue by adding a DriverLauncher which invokes java using jni

JIRA: [REEF-642](https://issues.apache.org/jira/browse/REEF-642)

This closes #

### CODE REVIEW

**Comment by markusweimer** [comment]
I'll do a pass.

File: lang/cs/.nuget/NuGet.Config [hunk 1]
```diff
@@ -21,11 +21,7 @@ under the License.
   <solution>
     <add key="disableSourceControlIntegration" value="true" />
   </solution>
-  <packageSources>
```
Comment (markusweimer) [inline lang/cs/.nuget/NuGet.Config]: These changes seem orthogonal to the PR. Please remove.

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 2]
```diff
@@ -0,0 +1,353 @@
...
+		return failureCode;
+	}
+
+	for (int i = 0; i < optionCount; i++) {
+		printf("Option %d [%s]\n", i, options[i].optionString);
+	}
+	fflush(stdout);
+	JavaVMInitArgs vm_args;
+	memset(&vm_args, 0, sizeof(vm_args));
+	vm_args.version = JNI_VERSION_1_6;
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: Does this tie us to Java 1.6?

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs [hunk 3]
```diff
@@ -38,13 +38,23 @@ internal sealed class DriverFolderPreparationHelper
     {
         private const string DLLFileNameExtension = ".dll";
         private const string EXEFileNameExtension = ".exe";
+        private const string BridgeExe = "Org.Apache.REEF.Bridge.exe";
+        private const string BridgeExeConfig = "Org.Apache.REEF.Bridge.exe.config";
+        private readonly string DefaultDriverConfigurationFileContents =
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: Could this be a `const`?

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs [hunk 4]
```diff
@@ -38,13 +38,23 @@ internal sealed class DriverFolderPreparationHelper
...
+        @"</configuration>";
+
         
         private readonly static Tuple<string, string>[] clientFileResources = new Tuple<string, string>[]
         {
             new Tuple<string, string>("ClientJarFullName", "reef_bridge_client"),
             new Tuple<string, string>("DriverJarFullName", "reef_bridge_driver"),
-            // enable with next pull request
-            // new Tuple<string, string>("ClrDriverFullName", "reef_clrdriver"),
+            new Tuple<string, string>("ClrDriverFullName", "reef_clrdriver"),
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: What do these strings signify? If they are magic strings, maybe they should be documented `consts`?

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs [hunk 5]
```diff
@@ -112,25 +122,39 @@ internal void CreateDriverConfiguration(IJobSubmission jobSubmission, string dri
...
-        /// Adds all Assemlies to the Global folder in the Driver.
-        /// </summary>
-        private void AddAssemblies()
-        {
             var resourceHelper = new ResourceHelper(typeof(DriverFolderPreparationHelper).Assembly);
             foreach (var fileResources in clientFileResources)
             {
                 File.WriteAllBytes(resourceHelper.GetString(fileResources.Item1), resourceHelper.GetBytes(fileResources.Item2));
             }
+            var exePathLocal = Path.Combine(driverFolderPath, BridgeExe);
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: Why are the two paths different?

I'm concerned that this might create debugging issues of the kind where it works on local, but doesn't work on YARN. Also, it might make porting to other RMs harder.

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/AllocatedEvaluatorBridge.java [hunk 6]
```diff
@@ -18,6 +18,7 @@
  */
 package org.apache.reef.javabridge;
 
+import org.apache.reef.io.naming.Identifiable;
```
Comment (markusweimer) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/AllocatedEvaluatorBridge.java]: This change is a positive, but it seems unrelated to the JIRA at hand. Maybe do it separately? Or at least file a JIRA and we can resolve that one with a link to this PR.

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+const char* NativeInteropClass = "org/apache/reef/javabridge/NativeInterop";
+const int maxPathBufSize = 16 * 1024;
+const char ClassPathSeparatorCharForWindows = ';';
+
+const char* launcherClass = "org.apache.reef.runtime.common.REEFLauncher";
+const char* JavaMainMethodName = "main";
+const char* JavaClassPath = "-Djava.class.path=";
+
+
+typedef enum {
```
Comment (markusweimer) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: Will these every show up in error prints? If so, it would be good to have some documentation here.

**Comment by markusweimer** [comment]
I've done a pass. I have not looked at the C++ code in detail, though. The code builds and passes RAT. However, I had five test failures:

```
Test method Org.Apache.REEF.Tests.Functional.Driver.TestDriver.TestDriverStart threw exception: 

System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out
    at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in Exceptions.cs: line 54
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in Exceptions.cs: line 72
   at Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper.GetAssembliesListForReefDriverApp() in ClrHandlerHelper.cs: line 130
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.UpdateJarFileWithAssemblies(String reefJar) in ClrClientHelper.cs: line 89
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.Run(HashSet`1 appDlls, IConfiguration driverBridgeConfig, DriverSubmissionSettings driverSubmissionSettings, String reefJar, String runCommand, String clrFolder, String className) in ClrClientHelper.cs: line 42
   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.TestRun(HashSet`1 appDlls, IConfiguration driverBridgeConfig, Boolean runOnYarn, JavaLoggingSetting javaLogSettings) in ReefFunctionalTest.cs: line 119
   at Org.Apache.REEF.Tests.Functional.Driver.TestDriver.TestDriverStart() in TestDriver.cs: line 64
```

```
Test method Org.Apache.REEF.Tests.Functional.Group.BroadcastReduceTest.TestBroadcastAndReduceOnLocalRuntime threw exception: 

System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out
    at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in Exceptions.cs: line 54
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in Exceptions.cs: line 72
   at Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper.GetAssembliesListForReefDriverApp() in ClrHandlerHelper.cs: line 130
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.UpdateJarFileWithAssemblies(String reefJar) in ClrClientHelper.cs: line 89
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.Run(HashSet`1 appDlls, IConfiguration driverBridgeConfig, DriverSubmissionSettings driverSubmissionSettings, String reefJar, String runCommand, String clrFolder, String className) in ClrClientHelper.cs: line 42
   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.TestRun(HashSet`1 appDlls, IConfiguration driverBridgeConfig, Boolean runOnYarn, JavaLoggingSetting javaLogSettings) in ReefFunctionalTest.cs: line 119
   at Org.Apache.REEF.Tests.Functional.Group.BroadcastReduceTest.TestBroadcastAndReduce(Boolean runOnYarn, Int32 numTasks) in BroadcastReduceTest.cs: line 114
   at Org.Apache.REEF.Tests.Functional.Group.BroadcastReduceTest.TestBroadcastAndReduceOnLocalRuntime() in BroadcastReduceTest.cs: line 59
```

```
Test method Org.Apache.REEF.Tests.Functional.Group.PipelinedBroadcastReduceTest.TestPipelinedBroadcastAndReduceOnLocalRuntime threw exception: 

System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out
    at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in Exceptions.cs: line 54
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in Exceptions.cs: line 72
   at Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper.GetAssembliesListForReefDriverApp() in ClrHandlerHelper.cs: line 130
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.UpdateJarFileWithAssemblies(String reefJar) in ClrClientHelper.cs: line 89
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.Run(HashSet`1 appDlls, IConfiguration driverBridgeConfig, DriverSubmissionSettings driverSubmissionSettings, String reefJar, String runCommand, String clrFolder, String className) in ClrClientHelper.cs: line 42
   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.TestRun(HashSet`1 appDlls, IConfiguration driverBridgeConfig, Boolean runOnYarn, JavaLoggingSetting javaLogSettings) in ReefFunctionalTest.cs: line 119
   at Org.Apache.REEF.Tests.Functional.Group.PipelinedBroadcastReduceTest.TestBroadcastAndReduce(Boolean runOnYarn, Int32 numTasks) in PipelinedBroadcastReduceTest.cs: line 116
   at Org.Apache.REEF.Tests.Functional.Group.PipelinedBroadcastReduceTest.TestPipelinedBroadcastAndReduceOnLocalRuntime() in PipelinedBroadcastReduceTest.cs: line 59
```

```
Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper Start: 0 : 2015-09-09T13:29:43.9787643-07:00 0015
START: 2015-09-09 13:29:43 ClrHandlerHelper::CopyDllsToAppDirectory
Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper Stop: 0 : 2015-09-09T13:29:43.9847512-07:00 0015
EXIT: 2015-09-09 13:29:43 ClrHandlerHelper::CopyDllsToAppDirectory. Duration: [00:00:00.0061807].
Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper Start: 0 : 2015-09-09T13:29:43.9867607-07:00 0015
START: 2015-09-09 13:29:43 ClrHandlerHelper::GetAssembliesListForReefDriverApp
Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper Error: 0 : 2015-09-09T13:29:44.0277747-07:00 0015
ERROR: ExceptionThrowing InvalidOperationException encountered error [System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out] with mesage [Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out] and stack trace []
Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper Stop: 0 : 2015-09-09T13:29:44.0287733-07:00 0015
EXIT: 2015-09-09 13:29:43 ClrHandlerHelper::GetAssembliesListForReefDriverApp. Duration: [00:00:00.0413319].
Cleaning up test.
Test method Org.Apache.REEF.Tests.Functional.Group.ScatterReduceTest.TestScatterAndReduceOnLocalRuntime threw exception: 

System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out
    at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in Exceptions.cs: line 54
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in Exceptions.cs: line 72
   at Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper.GetAssembliesListForReefDriverApp() in ClrHandlerHelper.cs: line 130
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.UpdateJarFileWithAssemblies(String reefJar) in ClrClientHelper.cs: line 89
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.Run(HashSet`1 appDlls, IConfiguration driverBridgeConfig, DriverSubmissionSettings driverSubmissionSettings, String reefJar, String runCommand, String clrFolder, String className) in ClrClientHelper.cs: line 42
   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.TestRun(HashSet`1 appDlls, IConfiguration driverBridgeConfig, Boolean runOnYarn, JavaLoggingSetting javaLogSettings) in ReefFunctionalTest.cs: line 119
   at Org.Apache.REEF.Tests.Functional.Group.ScatterReduceTest.TestScatterAndReduce(Boolean runOnYarn, Int32 numTasks) in ScatterReduceTest.cs: line 111
   at Org.Apache.REEF.Tests.Functional.Group.ScatterReduceTest.TestScatterAndReduceOnLocalRuntime() in ScatterReduceTest.cs: line 59
```

```
Test method Org.Apache.REEF.Tests.Functional.Messaging.TestTaskMessage.TestSendTaskMessage threw exception: 

System.InvalidOperationException: Assembly [Org.Apache.REEF.Bridge.dll] needed for REEF driver not found in C:\src\reef\lang\cs\TestResults\Deploy_mweimer 2015-09-09 13_29_40\Out
    at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, String message, Logger logger) in Exceptions.cs: line 54
   at Org.Apache.REEF.Utilities.Diagnostics.Exceptions.Throw(Exception exception, Logger logger) in Exceptions.cs: line 72
   at Org.Apache.REEF.Driver.Bridge.ClrHandlerHelper.GetAssembliesListForReefDriverApp() in ClrHandlerHelper.cs: line 130
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.UpdateJarFileWithAssemblies(String reefJar) in ClrClientHelper.cs: line 89
   at Org.Apache.REEF.Driver.Bridge.ClrClientHelper.Run(HashSet`1 appDlls, IConfiguration driverBridgeConfig, DriverSubmissionSettings driverSubmissionSettings, String reefJar, String runCommand, String clrFolder, String className) in ClrClientHelper.cs: line 42
   at Org.Apache.REEF.Tests.Functional.ReefFunctionalTest.TestRun(HashSet`1 appDlls, IConfiguration driverBridgeConfig, Boolean runOnYarn, JavaLoggingSetting javaLogSettings) in ReefFunctionalTest.cs: line 119
   at Org.Apache.REEF.Tests.Functional.Messaging.TestTaskMessage.TestSendTaskMessage() in TestTaskMessage.cs: line 80
```

Also, I wonder whether we could attack [REEF-383](https://issues.apache.org/jira/browse/REEF-383) as part of this or whether we should do that separately. Once this change is in, the likely solution to REEF-383 seems to be to delete `JavaBridge` altogether.

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs [hunk 8]
```diff
@@ -38,13 +38,23 @@ internal sealed class DriverFolderPreparationHelper
...
+        private const string BridgeExeConfig = "Org.Apache.REEF.Bridge.exe.config";
+        private readonly string DefaultDriverConfigurationFileContents =
+        @"<configuration>" +
+        @"  <runtime>" +
+        @"    <assemblyBinding xmlns=""urn:schemas-microsoft-com:asm.v1"">" +
+        @"      <probing privatePath=""reef/local;reef/global;local;global""/>" +
+        @"    </assemblyBinding>" +
+        @"  </runtime>" +
+        @"</configuration>";
+
```
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: Please add:

<gcServer enabled="true" />

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 2 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: jni and java versions are somewhat different. If the java version bump does not impact jni, they do not touch it.

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs [hunk 9]
```diff
@@ -112,25 +122,39 @@ internal void CreateDriverConfiguration(IJobSubmission jobSubmission, string dri
...
+            var exePathYarn = Path.Combine(driverFolderPath, _fileNames.GetReefFolderName(), BridgeExe);
+            var configPathLocal = Path.Combine(driverFolderPath, BridgeExeConfig);
+            var configPathYarn = Path.Combine(driverFolderPath, _fileNames.GetReefFolderName(), BridgeExeConfig);
+            File.Copy(BridgeExe, exePathLocal);
+            File.Copy(BridgeExe, exePathYarn);
+            var config = DefaultDriverConfigurationFileContents;
+            if (!string.IsNullOrEmpty(jobSubmission.DriverConfigurationFileContents))
+            {
+                config = jobSubmission.DriverConfigurationFileContents;
+            }
```
Comment (anupam128) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: what happens if user provided string does not include "probing privatePath=foo"?

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs (hunk 3 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: fixed

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs (hunk 4 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: fixed

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs (hunk 5 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: Unified

File: lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/AllocatedEvaluatorBridge.java (hunk 6 above)
Comment (beysims) [inline lang/java/reef-bridge-java/src/main/java/org/apache/reef/javabridge/AllocatedEvaluatorBridge.java]: created https://issues.apache.org/jira/browse/REEF-734

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 7 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: Added comments.

**Comment by beysims** [comment]
Added https://issues.apache.org/jira/browse/REEF-735 for test failures. 
@markusweimer  I am done for this iteration

**Comment by markusweimer** [comment]
What will we do about the failing tests in this PR? For we need to address REEF-735 before merging this?

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+*
+* Unless required by applicable law or agreed to in writing,
+* software distributed under the License is distributed on an
+* "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+* KIND, either express or implied.  See the License for the
+* specific language governing permissions and limitations
+* under the License.
+*/
+#include "InteropUtil.h"
+using namespace std;;
```
Comment (yingdachen) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: redundant ";"

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+	}
+}
+
+//
+// Set JVM option. JNI does not support unicode
+//
+void SetOption(JavaVMOption *option, char *src)
+{
+	size_t len = strlen(src) + 1;
+	char* pszOption = new char[len];
```
Comment (yingdachen) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: no delete, memory leak

in general i would suggest that we use std::string instead of raw chars

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 12]
```diff
@@ -0,0 +1,376 @@
...
+	array<String^>^ rawPaths = pathString->Split(';');
+	for (int i = 0; i < rawPaths->Length; i++)
+	{
+		String^ oldPath = rawPaths[i];
+		int oldPathLength = oldPath->Length;
+		String^ path;
+		bool shouldExpand = false;
+		if (oldPath->EndsWith("*"))
+		{
+			path = oldPath + ".jar";
```
Comment (yingdachen) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: nit: make ".jar" a constant String^ ?

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 13]
```diff
@@ -0,0 +1,376 @@
...
+	option->extraInfo = nullptr;
+}
+
+// 
+// Jni does not expand * and *.jar 
+// We need to do it ourselves.
+// 
+char *ExpandJarPaths(char *jarPaths)
+{
+	String^ ClassPathSeparatorStringForWindows = ";";
```
Comment (yingdachen) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: to be consistent, classPathSeparatorStringForWindows

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 11 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: The lifetime of the option allocation is same as lifetime of the executable.

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 12 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: fixed

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 13 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: fixed

File: lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs (hunk 9 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/Common/DriverFolderPreparationHelper.cs]: Created https://issues.apache.org/jira/browse/REEF-740

**Comment by beysims** [comment]
@markusweimer 
@anupam128 
@yingdachen 
I am done with this iteration

**Comment by jwang98052** [comment]
I will do a pass.

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs [hunk 14]
```diff
@@ -61,5 +61,12 @@ public interface IJobSubmission
         /// The Job's identifier
         /// </summary>
         string JobIdentifier { get; }
+
+        /// <summary>
+        /// Driver config file contents (Org.Apache.REEF.Bridge.exe.config) contents
+        /// Can be use to redirect assembly versions
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs]: used

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp [hunk 15]
```diff
@@ -0,0 +1,381 @@
...
+	return options;
+}
+
+int Get_CreateJavaVM_Function(JNI_CreateJavaVM_FN& fn_JNI_CreateJavaVM)
+{
+	wchar_t jvmDllPath1[maxPathBufSize];
+	wchar_t jvmDllPath2[maxPathBufSize];
+	DWORD rc = GetEnvironmentVariable(JAVA_HOME, jvmDllPath1, maxPathBufSize);
+	if (0 == rc) {
+		wprintf(L"Could not GetEnvironmentVariable %ls\n", JAVA_HOME);
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: A tab should be 4 chars. It is not that important though.

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs [hunk 16]
```diff
@@ -76,5 +76,13 @@ public interface IJobSubmissionBuilder
         /// Set driver memory in megabytes
         /// </summary>
         IJobSubmissionBuilder SetDriverMemory(int driverMemoryInMb);
+
+        /// <summary>
+        /// Driver config file contents (Org.Apache.REEF.Bridge.exe.config) contents
+        /// Can be use to redirect assembly versions
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs]: used

File: lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs [hunk 17]
```diff
@@ -134,8 +134,7 @@ private static Boolean IsAssembly(string path)
             {
                 return false;
             }
-            var extension = Path.GetExtension(path).ToLower();
-            return (extension == ".dll" || extension == ".exe") && (BinaryType.Clr == ((BinaryType)IsManagedBinary(path)));
+            return Path.GetExtension(path).ToLower() == ".dll";
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs]: Is IsAssembly() method here still used?

File: lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs [hunk 18]
```diff
@@ -110,11 +110,11 @@ private ISet<string> GetAssembliesInGlobalFolder()
...
-                .Where(IsAssembly)
+                .Where(Org.Apache.REEF.Tang.Util.AssemblyLoader.IsAssembly)
                 .Select(Path.GetFileNameWithoutExtension));
         }
 
-        [DllImport("Org.Apache.REEF.Bridge.dll", CharSet = CharSet.Unicode, SetLastError = true)]
+        [DllImport("Org.Apache.REEF.Bridge.exe", CharSet = CharSet.Unicode, SetLastError = true)]
         static extern int IsManagedBinary(string lpFileName);
 
         private enum BinaryType
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs]: Not used any more?

File: lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj [hunk 19]
```diff
@@ -42,14 +42,14 @@ under the License.
   </PropertyGroup>
   <Import Project="$(VCTargetsPath)\Microsoft.Cpp.Default.props" />
   <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='Debug|x64'" Label="Configuration">
-    <ConfigurationType>DynamicLibrary</ConfigurationType>
+    <ConfigurationType>Application</ConfigurationType>
```
Comment (jwang98052) [inline lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj]: move it to the common PropertyGroup

**Comment by jwang98052** [comment]
There are test failures in java reef- common:

Tests in error:
  JavaLaunchCommandBuilderTest.testOverrideMemory:99 ╗ NullPointer
  JavaLaunchCommandBuilderTest.testAddMemoryOption:108 ╗ NullPointer
  JavaLaunchCommandBuilderTest.testAddPermSizeOptions:143 ╗ NullPointer
  JavaLaunchCommandBuilderTest.testPermSizeDefaults:136 ╗ NullPointer
  JavaLaunchCommandBuilderTest.testOverrideMemoryOption:116 ╗ NullPointer
  JavaLaunchCommandBuilderTest.testSetMemory:91 ╗ NullPointer


[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary:
[INFO]
[INFO] REEF ............................................... SUCCESS [  6.524 s]
[INFO] REEF Dot Net ....................................... SUCCESS [  2.796 s]
[INFO] REEF Annotations ................................... SUCCESS [  1.970 s]
[INFO] REEF Utils ......................................... SUCCESS [ 18.976 s]
[INFO] REEF Tang Project .................................. SUCCESS [  0.604 s]
[INFO] REEF Tang .......................................... SUCCESS [ 15.524 s]
[INFO] REEF Wake Project .................................. SUCCESS [  0.573 s]
[INFO] REEF Wake .......................................... SUCCESS [01:20 min]
[INFO] REEF Common ........................................ FAILURE [ 26.380 s]
[INFO] REEF Runtime Local ................................. SKIPPED
[INFO] REEF Applications .................................. SKIPPED
[INFO] REEF Vortex ........................................ SKIPPED
[INFO] REEF Utils for Hadoop .............................. SKIPPED

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs (hunk 14 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmission.cs]: fixed

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 15 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: Fixed

File: lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs (hunk 16 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Client/API/IJobSubmissionBuilder.cs]: Fixed

File: lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs (hunk 17 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs]: Removed

File: lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs (hunk 18 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Driver/ClassHierarchyGeneratingDriverStartObserver.cs]: Removed

File: lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj (hunk 19 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/Org.Apache.REEF.Bridge.vcxproj]: Moved.

**Comment by beysims** [comment]
Fixed.

**Comment by beysims** [comment]
@markusweimer 
@jwang98052 
I am done with this iteration

File: lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp (hunk 10 above)
Comment (beysims) [inline lang/cs/Org.Apache.REEF.Bridge/DriverLauncher.cpp]: fixed

**Comment by jwang98052** [comment]
testing it now

**Comment by jwang98052** [comment]
Java test passed, .Net tests passed. I will merge it.
